src/hip_module.cpp

```diff
@@ -459,89 +542,133 @@ hipError_t ihipLaunchCooperativeKernel(const void* f, dim3 gridDim,
-#endif
+    coopAV.acquire_locked_hsa_queue();
 
-    GET_TLS();
-    // launch the init_gws kernel to initialize the GWS
+    // launch the init_gws kernel to initialize the GWS in the dedicated cooperative queue
     result = ihipModuleLaunchKernel(tls, gwsKD, 1, 1, 1, 1, 1, 1,
-             0, stream, gwsKernelParam, nullptr, nullptr, nullptr, 0, true);
+             0, stream, gwsKernelParam, nullptr, nullptr, nullptr, 0, true, nullptr , &coopAV);
 
     if (result != hipSuccess) {
         stream->criticalData().unlock();
-#if (__hcc_workweek__ >= 19213)
         stream->criticalData()._av.release_locked_hsa_queue();
-#endif
-
+        coopAV.release_locked_hsa_queue();
         return hipErrorLaunchFailure;
     }
 
     size_t impCoopArg = 1;
     void* impCoopParams[1];
     impCoopParams[0] = &impCoopArg;
 
-    // launch the main kernel
+    // launch the main kernel in the cooperative queue
     result = ihipModuleLaunchKernel(tls, kd,
             gridDim.x * blockDimX.x,
             gridDim.y * blockDimX.y,
             gridDim.z * blockDimX.z,
             blockDimX.x, blockDimX.y, blockDimX.z,
             sharedMemBytes, stream, kernelParams, nullptr, nullptr,
-            nullptr, 0, true, impCoopParams);
+            nullptr, 0, true, impCoopParams, &coopAV);
 
     stream->criticalData().unlock();
-#if (__hcc_workweek__ >= 19213)
     stream->criticalData()._av.release_locked_hsa_queue();
-#endif
+    coopAV.release_locked_hsa_queue();
+
+    // wait on the dispatch on the dedicated cooperative queue to finish
+    coopAV.wait(hc::hcWaitModeActive);
+
 
     return result;
+#else
+    return hipErrorInvalidConfiguration;
+#endif
+
 }
 
 __attribute__((visibility("default")))
 hipError_t ihipLaunchCooperativeKernelMultiDevice(hipLaunchParams* launchParamsList,
         int  numDevices, unsigned int  flags, hip_impl::program_state& ps) {
 
+#if (__hcc_workweek__ >= 20093)
     hipError_t result;
 
     if (numDevices > g_deviceCnt || launchParamsList == nullptr || numDevices > MAX_COOPERATIVE_GPUs) {
         return hipErrorInvalidValue;
     }
 
+    vector<hipStream_t> streams;
+    vector<uint64_t> deviceIDs;
+    // check to see if we have valid distinct streams/devices, if cooperative multi device
+    // launch is supported and if grid/block dimensions are valid
     for (int i = 0; i < numDevices; ++i) {
-        if (!launchParamsList[i].stream->getDevice()->_props.cooperativeMultiDeviceLaunch) {
-            return hipErrorInvalidConfiguration;
+        const hipLaunchParams& lp = launchParamsList[i];
+
+        if (lp.stream == nullptr){
+            return hipErrorInvalidResourceHandle;
         }
-    }
 
-    hipFunction_t* gwsKds = reinterpret_cast<hipFunction_t*>(malloc(sizeof(hipFunction_t) * numDevices));
-    hipFunction_t* kds    = reinterpret_cast<hipFunction_t*>(malloc(sizeof(hipFunction_t) * numDevices));
+        auto it = find(streams.begin(), streams.end(), lp.stream);
+        if (it == streams.end()){
```

Comment:
Here and below iterator is excessive, `if (find(...) == steams. end()) `.

Reply:
corrected.